src/Build/Collections/CopyOnReadEnumerable.cs

```diff
@@ -46,13 +56,23 @@ public CopyOnReadEnumerable(IEnumerable<T> backingEnumerable, object syncRoot)
         /// Returns an enumerator over the collection.
         /// </summary>
         /// <returns>The enumerator.</returns>
-        public IEnumerator<T> GetEnumerator()
+        public IEnumerator<TResult> GetEnumerator()
         {
-            List<T> list;
+            List<TResult> list;
+            if (_backingEnumerable is ICollection backingCollection)
+            {
+                list = new List<TResult>(backingCollection.Count);
```

Comment:
Should this use [TryGetNonEnumeratedCount](https://docs.microsoft.com/en-us/dotnet/api/system.linq.enumerable.trygetnonenumeratedcount?view=net-6.0)?

Reply:
If there is net6.0 multi-targeting support.

Reply:
Nice, I didn't know about this method, thank you. I've added an #ifdef to make it super easy to delete the old code when we remove Framework support in the future.